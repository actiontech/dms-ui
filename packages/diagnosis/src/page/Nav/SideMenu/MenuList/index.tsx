import { MenuProps, Menu } from 'antd5';
import { SubMenuType } from 'antd/lib/menu/hooks/useItems';
import { useMemo, useCallback } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { MenuListProps } from './index.type';
import { sideMenuData } from '../menus/menu.data';

const MenuList: React.FC<MenuListProps> = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const menuItems = useMemo(() => sideMenuData(navigate), [navigate]);

  const selectMenu = useCallback(
    (config: MenuProps['items'] = [], pathname: string): string[] => {
      for (const route of config) {
        if (!route) {
          return [];
        }
        if (route.key === pathname) {
          return [(route.key as string) ?? ''];
        }
        if (route) {
          const children = (route as SubMenuType).children;
          if (children) {
            const keys = selectMenu(children, pathname);
            if (keys.length > 0) {
              return keys;
            }
          }
        }
      }
      return [];
    },
    []
  );

  const selectMenuWrapper = useMemo((): string[] => {
    let pathname = location.pathname;
    let selectKey: string[] = [];
    while (pathname.length > 0) {
      selectKey = selectMenu(menuItems, pathname);
      if (selectKey.length !== 0) {
        return selectKey;
      } else {
        const temp = pathname.split('/');
        temp.pop();
        pathname = temp.join('/');
      }
    }
    return selectKey;
  }, [location.pathname, menuItems, selectMenu]);
  console.log(menuItems);
  return (
    <Menu
      className="custom-menu"
      mode="inline"
      items={menuItems}
      selectedKeys={selectMenuWrapper}
    />
  );
};

export default MenuList;
