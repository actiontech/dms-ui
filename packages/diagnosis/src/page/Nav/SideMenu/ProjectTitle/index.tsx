import { useNavigate } from 'react-router-dom';

const ProjectTitle: React.FC = () => {
  const navigate = useNavigate();

  return (
    <div
      onClick={() => {
        navigate('/');
      }}
    >
      <div className="title">
        <span className="label label-primary">Action</span>
        <span className="label label-base">Diagnosis</span>
      </div>
    </div>
  );
};

export default ProjectTitle;
