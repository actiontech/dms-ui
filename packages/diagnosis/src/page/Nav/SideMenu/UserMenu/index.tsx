import UserNavigate from './UserNavigate';
import { UserMenuProps } from './index.type';

const UserMenu: React.FC<UserMenuProps> = ({ username }) => {
  return (
    <>
      <div className="dms-layout-side-end">
        <UserNavigate username={username} />
      </div>
    </>
  );
};

export default UserMenu;
