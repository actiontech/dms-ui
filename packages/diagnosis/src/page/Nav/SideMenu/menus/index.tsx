import i18n from 'i18next';
import { GenerateMenuItemsType } from './common';
import { ThunderBulbFilled, UserShieldFilled } from '@actiontech/icons';

export const DiagnosisMenuItems: GenerateMenuItemsType = ({ navigate }) => [
  {
    order: 0,
    type: 'divider'
  },
  {
    order: 1,
    label: <>{i18n.t('diagnosisMenu.monitorSourceConfig')}</>,
    key: `/monitor-source-config`,
    icon: <ThunderBulbFilled width={18} height={18} />,
    onClick: () => navigate(`/monitor-source-config`)
  },
  {
    order: 2,
    label: <>{i18n.t('diagnosisMenu.userManagement')}</>,
    key: `/user-management`,
    icon: <UserShieldFilled width={18} height={18} />,
    onClick: () => navigate(`/user-management`)
  }
];
