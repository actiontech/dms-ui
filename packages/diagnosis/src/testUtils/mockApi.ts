import { AxiosResponse } from 'axios';

// some api return data params is not data, like V1GetServerHostname
export const createSpySuccessResponseWithoutDataParams = (
  data: any,
  { status = 200, headers = {}, config = {}, statusText = '' } = {}
) => {
  return new Promise<AxiosResponse<any>>((res) => {
    setTimeout(() => {
      res({
        status,
        headers,
        config,
        statusText,
        data: {
          code: 0,
          msg: 'ok',
          ...data
        }
      });
    }, 3000);
  });
};
