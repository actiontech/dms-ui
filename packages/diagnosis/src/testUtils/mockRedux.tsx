import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { Dictionary } from '@actiontech/shared/lib/types/common.type';
import monitorSourceConfig from '../store/monitorSourceConfig';
import user from '../store/user';
import userManagement from '../store/userManagement';

const reducers = combineReducers({
  user,
  monitorSourceConfig,
  userManagement
});

export const storeFactory = (initStore: Dictionary = {}) => {
  return configureStore({ reducer: reducers, preloadedState: initStore });
};
