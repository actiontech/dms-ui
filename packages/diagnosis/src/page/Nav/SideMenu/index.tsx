import { SideMenuStyleWrapper } from './style';
import { useCurrentUser } from '@actiontech/shared/lib/global';
import UserMenu from './UserMenu';
import ProjectTitle from './ProjectTitle';
import MenuList from './MenuList';

const SideMenu: React.FC = () => {
  const { username, theme, updateTheme, isAdmin } = useCurrentUser();

  return (
    <SideMenuStyleWrapper className="diagnosis-layout-side">
      <div className="diagnosis-layout-side-start">
        <ProjectTitle />
        <MenuList isAdmin={isAdmin} />
      </div>
      <UserMenu
        username={username}
        updateTheme={updateTheme}
        isAdmin={isAdmin}
        theme={theme}
      />
    </SideMenuStyleWrapper>
  );
};

export default SideMenu;
