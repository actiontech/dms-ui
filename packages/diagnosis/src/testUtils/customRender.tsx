import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter, MemoryRouterProps } from 'react-router-dom';
import { Dictionary } from '@actiontech/shared/lib/types/common.type';
import { storeFactory } from './mockRedux';
import { StyledEngineProvider, ThemeProvider } from '@mui/system';
import { mount } from 'enzyme';
import { ConfigProvider } from 'antd';
import zhCN from 'antd/locale/zh_CN';

import lightTheme from '../theme/light';
import sharedTheme from '@actiontech/shared/lib/theme/light';

type MountParams = Parameters<typeof mount>;
type RenderParams = Parameters<typeof render>;

export const renderWithReduxAndTheme = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(ui, {
    wrapper: ({ children }) => (
      <ThemeProvider theme={{ ...lightTheme, ...sharedTheme }}>
        <Provider store={storeFactory(initStore)}>{children}</Provider>
      </ThemeProvider>
    ),
    ...Option
  });
};

export const renderWithTheme = (...[ui, option]: [...RenderParams]) => {
  return render(
    <ThemeProvider theme={{ ...lightTheme, ...sharedTheme }}>
      {ui}
    </ThemeProvider>,
    option
  );
};

export const renderWithThemeAndRouter = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(
    <MemoryRouter {...props}>
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={{ ...lightTheme, ...sharedTheme }}>
          {ui}
        </ThemeProvider>
      </StyledEngineProvider>
    </MemoryRouter>,
    option
  );
};

export const superRender = (
  ...[ui, option, otherProps]: [
    ...RenderParams,
    {
      routerProps?: MemoryRouterProps;
      initStore?: any;
    }?
  ]
) => {
  const renderReturn = render(ui, {
    wrapper: ({ children }) => {
      return (
        <ConfigProvider locale={zhCN}>
          <ThemeProvider theme={{ ...lightTheme, ...sharedTheme }}>
            <MemoryRouter {...otherProps?.routerProps}>
              <Provider store={storeFactory(otherProps?.initStore)}>
                {children}
              </Provider>
            </MemoryRouter>
          </ThemeProvider>
        </ConfigProvider>
      );
    },
    ...option
  });
  return renderReturn;
};

export const mountWithTheme = (...[ui, option]: [...MountParams]) => {
  return mount(ui, {
    ...option,
    wrappingComponent: ThemeProvider,
    wrappingComponentProps: {
      theme: { ...lightTheme, ...sharedTheme }
    }
  });
};
