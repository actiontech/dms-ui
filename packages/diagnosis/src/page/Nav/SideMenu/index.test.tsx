import SideMenu from '.';
import { superRender } from '../../../testUtils/customRender';
import { screen, act } from '@testing-library/react';

describe('diagnosis/SideMenu', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should match snapshot', async () => {
    const { baseElement } = superRender(<SideMenu />);
    await act(async () => {
      jest.runOnlyPendingTimers();
    });
    expect(baseElement).toMatchSnapshot();
    expect(screen.getByText('Action')).toBeInTheDocument();
    expect(screen.getByText('Diagnosis')).toBeInTheDocument();
  });
});
