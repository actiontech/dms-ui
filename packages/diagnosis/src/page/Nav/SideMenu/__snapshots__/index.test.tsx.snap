// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diagnosis/SideMenu click other menu 1`] = `
<body>
  <div>
    <div
      class="dms-layout-side css-2df5or"
    >
      <div
        class="dms-layout-side-start"
      >
        <div
          class="home-page-shortcut"
        >
          <div
            class="title"
          >
            <span
              class="label label-primary"
            >
              Action
            </span>
            <span
              class="label label-base"
            >
              Diagnosis
            </span>
          </div>
        </div>
        <ul
          class="ant-menu ant-menu-root ant-menu-inline ant-menu-light custom-menu css-dev-only-do-not-override-txh9fw"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-menu-item-divider"
          />
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-/monitor-source-config"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="anticon ant-menu-item-icon"
              role="img"
            >
              <svg
                fill="none"
                height="18"
                viewBox="0 0 18 18"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.95576 13.5C5.73293 12.5456 4.72809 11.7643 4.31546 11.2494C3.49234 10.2223 3 8.91862 3 7.5C3 4.18629 5.68629 1.5 9 1.5C12.3137 1.5 15 4.18629 15 7.5C15 8.91937 14.5071 10.2236 13.6832 11.251C13.2708 11.7653 12.2669 12.5457 12.0443 13.5H5.95576ZM12 15V15.75C12 16.5785 11.3285 17.25 10.5 17.25H7.5C6.67157 17.25 6 16.5785 6 15.75V15H12ZM9.75 7.5036V4.5L6.375 9.0036H8.25V12.0036L11.625 7.5036H9.75Z"
                  fill="#C3C6CD"
                />
                <path
                  d="M12 15V15.75C12 16.5784 11.3285 17.25 10.5 17.25H7.5C6.67157 17.25 6 16.5784 6 15.75V15H12Z"
                  fill="#8A8F99"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              监控源配置
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-/user-management"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="anticon ant-menu-item-icon"
              role="img"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.522 1.88408L8 0.666748L13.478 1.88408C13.626 1.91699 13.7584 1.99939 13.8533 2.11768C13.9482 2.23598 14 2.38309 14 2.53475V9.19275C14 9.85124 13.8373 10.4995 13.5266 11.0801C13.2158 11.6606 12.7666 12.1555 12.2187 12.5207L8 15.3334L3.78133 12.5207C3.23352 12.1556 2.78431 11.6608 2.47357 11.0804C2.16282 10.5 2.00016 9.85179 2 9.19341V2.53475C2.00003 2.38309 2.05176 2.23598 2.14666 2.11768C2.24156 1.99939 2.37396 1.91699 2.522 1.88408ZM8 7.33341C8.44203 7.33341 8.86595 7.15782 9.17851 6.84526C9.49107 6.5327 9.66667 6.10878 9.66667 5.66675C9.66667 5.22472 9.49107 4.8008 9.17851 4.48824C8.86595 4.17568 8.44203 4.00008 8 4.00008C7.55797 4.00008 7.13405 4.17568 6.82149 4.48824C6.50893 4.8008 6.33333 5.22472 6.33333 5.66675C6.33333 6.10878 6.50893 6.5327 6.82149 6.84526C7.13405 7.15782 7.55797 7.33341 8 7.33341ZM5.018 10.6667H10.982C10.9012 9.93235 10.5522 9.2536 10.0019 8.76056C9.45168 8.26753 8.73883 7.99489 8 7.99489C7.26117 7.99489 6.54832 8.26753 5.99805 8.76056C5.44779 9.2536 5.09881 9.93235 5.018 10.6667Z"
                  fill="#C3C6CD"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              用户管理
            </span>
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
      </div>
      <div
        class="dms-layout-side-end"
      >
        <span
          class="ant-avatar ant-avatar-circle action-avatar css-1wq79dc css-dev-only-do-not-override-txh9fw"
        >
          <span
            class="ant-avatar-string"
            style="transform: scale(1) translateX(-50%);"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`diagnosis/SideMenu should match snapshot 1`] = `
<body>
  <div>
    <div
      class="dms-layout-side css-2df5or"
    >
      <div
        class="dms-layout-side-start"
      >
        <div
          class="home-page-shortcut"
        >
          <div
            class="title"
          >
            <span
              class="label label-primary"
            >
              Action
            </span>
            <span
              class="label label-base"
            >
              Diagnosis
            </span>
          </div>
        </div>
        <ul
          class="ant-menu ant-menu-root ant-menu-inline ant-menu-light custom-menu css-dev-only-do-not-override-txh9fw"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-menu-item-divider"
          />
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-/monitor-source-config"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="anticon ant-menu-item-icon"
              role="img"
            >
              <svg
                fill="none"
                height="18"
                viewBox="0 0 18 18"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.95576 13.5C5.73293 12.5456 4.72809 11.7643 4.31546 11.2494C3.49234 10.2223 3 8.91862 3 7.5C3 4.18629 5.68629 1.5 9 1.5C12.3137 1.5 15 4.18629 15 7.5C15 8.91937 14.5071 10.2236 13.6832 11.251C13.2708 11.7653 12.2669 12.5457 12.0443 13.5H5.95576ZM12 15V15.75C12 16.5785 11.3285 17.25 10.5 17.25H7.5C6.67157 17.25 6 16.5785 6 15.75V15H12ZM9.75 7.5036V4.5L6.375 9.0036H8.25V12.0036L11.625 7.5036H9.75Z"
                  fill="#C3C6CD"
                />
                <path
                  d="M12 15V15.75C12 16.5784 11.3285 17.25 10.5 17.25H7.5C6.67157 17.25 6 16.5784 6 15.75V15H12Z"
                  fill="#8A8F99"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              监控源配置
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-/user-management"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              class="anticon ant-menu-item-icon"
              role="img"
            >
              <svg
                fill="none"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2.522 1.88408L8 0.666748L13.478 1.88408C13.626 1.91699 13.7584 1.99939 13.8533 2.11768C13.9482 2.23598 14 2.38309 14 2.53475V9.19275C14 9.85124 13.8373 10.4995 13.5266 11.0801C13.2158 11.6606 12.7666 12.1555 12.2187 12.5207L8 15.3334L3.78133 12.5207C3.23352 12.1556 2.78431 11.6608 2.47357 11.0804C2.16282 10.5 2.00016 9.85179 2 9.19341V2.53475C2.00003 2.38309 2.05176 2.23598 2.14666 2.11768C2.24156 1.99939 2.37396 1.91699 2.522 1.88408ZM8 7.33341C8.44203 7.33341 8.86595 7.15782 9.17851 6.84526C9.49107 6.5327 9.66667 6.10878 9.66667 5.66675C9.66667 5.22472 9.49107 4.8008 9.17851 4.48824C8.86595 4.17568 8.44203 4.00008 8 4.00008C7.55797 4.00008 7.13405 4.17568 6.82149 4.48824C6.50893 4.8008 6.33333 5.22472 6.33333 5.66675C6.33333 6.10878 6.50893 6.5327 6.82149 6.84526C7.13405 7.15782 7.55797 7.33341 8 7.33341ZM5.018 10.6667H10.982C10.9012 9.93235 10.5522 9.2536 10.0019 8.76056C9.45168 8.26753 8.73883 7.99489 8 7.99489C7.26117 7.99489 6.54832 8.26753 5.99805 8.76056C5.44779 9.2536 5.09881 9.93235 5.018 10.6667Z"
                  fill="#C3C6CD"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              用户管理
            </span>
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
      </div>
      <div
        class="dms-layout-side-end"
      >
        <span
          class="ant-avatar ant-avatar-circle action-avatar css-1wq79dc css-dev-only-do-not-override-txh9fw"
        >
          <span
            class="ant-avatar-string"
            style="transform: scale(1) translateX(-50%);"
          />
        </span>
      </div>
    </div>
  </div>
</body>
`;
